packages/datadog-plugin-aws-sdk/src/index.js

```diff
@@ -0,0 +1,109 @@
+'use strict'
+
+const Tags = require('opentracing').Tags
+const analyticsSampler = require('../../dd-trace/src/analytics_sampler')
+const awsHelpers = require('./helpers')
+
+function createWrapRequest (tracer, config) {
+  config = normalizeConfig(config)
+
+  return function wrapRequest (request) {
+    return function requestWithTrace (operation, params, cb) {
+      const serviceName = awsHelpers.normalizeServiceName(this)
+      const childOf = tracer.scope().active()
+      const baseTags = {
+        [Tags.SPAN_KIND]: 'client',
+        'span.type': 'http',
+        'service.name': config.service ? `${config.service}-${serviceName}` : `${tracer._service}-${serviceName}`,
+        'aws.agent': 'js-aws-sdk',
+        'aws.operation': operation,
+        'aws.region': this.config.region,
+        'aws.service': serviceName,
+        'aws.url': this.endpoint && this.endpoint.href,
+        'component': serviceName
+      }
+      let span
+
+      if (typeof cb === 'function') {
+        span = tracer.startSpan('aws.http', {
+          childOf,
+          tags: baseTags
+        })
+
+        analyticsSampler.sample(span, config.analytics)
+
+        // when passed a callback makeRequest will call `.send` on the AWS.Request
+        // before we have a chance to add our listeners, so we have to wrap
```

Comment:
Is it possible to wrap the Request prototype so that we get to run first?

Reply:
I don't quite follow, do you mean, literally shim the entire `AWS.Request.prototype` or do you mean bind the `request` object before calling it? 

```
const boundRequest = tracer.scope().bind(span, request)
return boundRequest.call(
  this,
  operation,
  params,
  awsHelpers.wrapCallback(tracer, span, cb, childOf, config, serviceName)
)
```

Or am i shimming, `AWS.request.prototype.send`? Not totally clear what you mean/what the goal is here?

Reply:
I mean shim the entire prototype. Then we wouldn't need 2 code paths since we'd be guaranteed to run first.

Reply:
Just to be clear, by entire prototype I mean the method(s) where we need to run first, not every method.

Reply:
i went a level lower to shim `AWS.request.prototype.send` and `AWS.request.prototype.promise`. This also uncovered to me that min version that can be instrumented here is 2.1.35 (instead of 2.1.0) as there is a breaking change `<2.1.35` in the way the emit'ing of events occurs. This is from late 2015 so i think it's still a very reasonable default min version.

https://github.com/aws/aws-sdk-js/pull/629

Otherwise moving down a level lower simplifies the code paths quite a bit as the instrumentation can now always attach listeners before `.send()` is invoked and the state machine is kicked off.

Reply:
I encountered a slight hiccup here which is that when merely starting a span on `validate`, and finishing a span on `complete` events , the active scope was not being correctly picked up by nested integrations in `aws.request`. Specifically, that meant the `http`/`tcp`/`dns` integrations were associating with the parent span of `aws.request` (which would be an express span in the example below, and not `aws.request`. Here's a snapshot of this behaviour

![incorrect_child_parent_relationship](https://cl.ly/0ceab6d2d58a/%255B5e01f0b3703104804e12adfbb28aba2f%255D_incorrect_child_parent_relationship.png)

I've been able to correct this is by explicitly starting the span in the shim and then wrapping the invocation of the original `AWS.request.send()` in the `aws.request` span's activated scope. Since this is now a level lower and wrapping the `AWS.request.send` method explicitly, which is what kicks off the state machine being run until it's complete (instead of the first iteration which was wrapping  `aws.Service.makeRequest` , which merely instantiated the state machine but did not yet kicking off it's run)  this should still ensure the start time of the span is correct. Here's a snapshot of the updated behaviour:

![correct_child_parent_relationship](https://cl.ly/baf1bd66cf33/%255B6d0d3a49dfb26746e6540cef7729ca5c%255D_correct_child_parent_relationship.png)

I've also added tests for this behavior as well as tests to ensue that any callbacks passed in are associated with the parent scope and not the scope of `aws.request`

Reply:
The new behavior is not correct as it seems every AWS request is a child of the previous request. To fix this, the scope when `send` was called should be restored in the callback.

Reply:
@rochdev my bad i included an older screenshot, i've since correct this behavior

updated: 

![nested properly](https://cl.ly/3bb4a6a2f92a/Image%2525202020-02-24%252520at%2525208.52.48%252520PM.png)

Here's the test app i used to generate this (it assumes localstack test docker container is running which should be listening on the correct host/port as long as the test env is set up appropriately)

https://gist.github.com/ericmustin/80636a955bb68f345059665580b936d2

Here's the LOC and also the test that demonstrates the callback is bound to the correct parent scope

LOC: https://github.com/DataDog/dd-trace-js/blob/486361f85e79d3f9870c5dfd2a9f3ae1cbb0fa6b/packages/datadog-plugin-aws-sdk/src/index.js#L34
Test: https://github.com/DataDog/dd-trace-js/blob/486361f85e79d3f9870c5dfd2a9f3ae1cbb0fa6b/packages/datadog-plugin-aws-sdk/test/index.spec.js#L1069

Again, my bad on the older screenshot, was rushing.